File: langs/csharp/src/VMx/State/AsyncResourceVM.cs
using VMx.Commands;
using VMx.Components;
using VMx.Services;

namespace VMx.State;

/// <summary>The discriminant for <see cref="AsyncResourceState{T}"/>.</summary>
public enum AsyncResourceStatus
{
    /// <summary>No load has produced a current value or error.</summary>
    Idle,
    /// <summary>A loader is currently in flight.</summary>
    Loading,
    /// <summary>The current loader completed with an accepted value.</summary>
    Ready,
    /// <summary>The current loader completed with a captured failure.</summary>
    Error,
}

/// <summary>Controls whether a previously accepted value remains visible during reload.</summary>
public enum AsyncResourceRetention
{
    /// <summary>Relinquish an accepted value when reload begins.</summary>
    DiscardPrevious,
    /// <summary>Keep an accepted value visible while reload runs or fails.</summary>
    RetainPrevious,
}

/// <summary>Immutable, validated snapshot of one asynchronous resource.</summary>
public sealed class AsyncResourceState<T>
{
    private AsyncResourceState(
        AsyncResourceStatus status,
        bool hasValue,
        T? value,
        Exception? error)
    {
        Status = status;
        HasValue = hasValue;
        Value = value;
        Error = error;
    }

    /// <summary>Gets the state discriminant.</summary>
    public AsyncResourceStatus Status { get; }
    /// <summary>Gets whether this snapshot carries an accepted value.</summary>
    public bool HasValue { get; }
    /// <summary>Gets the accepted value when <see cref="HasValue"/> is true.</summary>
    public T? Value { get; }
    /// <summary>Gets the current loader failure in <see cref="AsyncResourceStatus.Error"/>.</summary>
    public Exception? Error { get; }

    internal static AsyncResourceState<T> Idle() =>
        new(AsyncResourceStatus.Idle, false, default, null);

    internal static AsyncResourceState<T> Loading() =>
        new(AsyncResourceStatus.Loading, false, default, null);

    internal static AsyncResourceState<T> Loading(T value) =>
        new(AsyncResourceStatus.Loading, true, value, null);

    internal static AsyncResourceState<T> Ready(T value) =>
        new(AsyncResourceStatus.Ready, true, value, null);

    internal static AsyncResourceState<T> Failed(Exception error) =>
        new(AsyncResourceStatus.Error, false, default, error);

    internal static AsyncResourceState<T> Failed(T value, Exception error) =>
        new(AsyncResourceStatus.Error, true, value, error);
}

/// <summary>
/// Component viewmodel for one cancellable asynchronously acquired presentation value.
/// See spec/23-async-resource-vm.md and ADR-0100.
/// </summary>
public sealed class AsyncResourceVM<T> : ComponentVMBase
{
    private readonly object _resourceGate = new();
    private readonly Func<CancellationToken, Task<T>> _loader;
    private readonly AsyncResourceRetention _retention;
    private readonly Action<T>? _cleanupValue;
    private AsyncResourceState<T> _state = AsyncResourceState<T>.Idle();
    private AsyncResourceState<T> _stableState = AsyncResourceState<T>.Idle();
    private ResourceOperation? _operation;
    private long _operationIdentity;
    private bool _resourceDisposed;

    /// <summary>Creates an async resource VM with explicit services and loader.</summary>
    public AsyncResourceVM(
        string name,
        Func<CancellationToken, Task<T>> loader,
        IMessageHub hub,
        IDispatcher dispatcher,
        string hint = "",
        AsyncResourceRetention retention = AsyncResourceRetention.DiscardPrevious,
        Action<T>? cleanupValue = null)
        : base(name, hint, hub, dispatcher, null, null)
    {
        _loader = loader ?? throw new ArgumentNullException(nameof(loader));
        _retention = retention;
        _cleanupValue = cleanupValue;

        LoadCommand = AsyncRelayCommand.Builder()
            .Task(LoadAsync)
            .Predicate(CanLoad)
            .Build();
        ReloadCommand = AsyncRelayCommand.Builder()
            .Task(ReloadAsync)
            .Predicate(CanReload)
            .Build();
        CancelCommand = RelayCommand.Builder()
            .Task(Cancel)
            .Predicate(CanCancel)
            .Build();
    }

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>Gets the current immutable presentation snapshot.</summary>
    public AsyncResourceState<T> State
    {
        get
        {
            lock (_resourceGate) return _state;
        }
    }

    /// <summary>Gets the initial-load command.</summary>
    public AsyncRelayCommand LoadCommand { get; }
    /// <summary>Gets the retry/refresh command.</summary>
    public AsyncRelayCommand ReloadCommand { get; }
    /// <summary>Gets the active-operation cancellation command.</summary>
    public RelayCommand CancelCommand { get; }

    /// <summary>Starts the initial load when state is idle.</summary>
    public Task LoadAsync(CancellationToken cancellationToken = default) =>
        StartAsync(StartIntent.Load, cancellationToken);

    /// <summary>Starts or supersedes a reload when state is non-idle.</summary>
    public Task ReloadAsync(CancellationToken cancellationToken = default) =>
        StartAsync(StartIntent.Reload, cancellationToken);

    /// <summary>Cancels the current operation and restores its stable baseline.</summary>
    public void Cancel()
    {
        ResourceOperation? operation;
        lock (_resourceGate)
        {
            if (!CanCancelUnsafe() || _operation is null) return;
            operation = _operation;
        }

        operation.Cancel();
        LoadCommand.Cancel();
        ReloadCommand.Cancel();
    }

    /// <inheritdoc/>
    protected override void OnDispose()
    {
        ResourceOperation? operation;
        bool hasAccepted;
        T? accepted;
        lock (_resourceGate)
        {
            if (_resourceDisposed) return;
            _resourceDisposed = true;
            unchecked { _operationIdentity++; }
            operation = _operation;
            _operation = null;
            hasAccepted = TryGetValue(_stableState, out accepted);
            _stableState = AsyncResourceState<T>.Idle();
        }

        operation?.Cancel();
        LoadCommand.Cancel();
        ReloadCommand.Cancel();
        LoadCommand.Dispose();
        ReloadCommand.Dispose();
        CancelCommand.Dispose();
        if (hasAccepted) Cleanup(accepted!);
    }

    private bool CanLoad()
    {
        lock (_resourceGate)
            return !_resourceDisposed && _state.Status == AsyncResourceStatus.Idle;
    }

    private bool CanReload()
    {
        lock (_resourceGate)
            return !_resourceDisposed && _state.Status != AsyncResourceStatus.Idle;
    }

    private bool CanCancel()
    {
        lock (_resourceGate) return CanCancelUnsafe();
    }

    private bool CanCancelUnsafe() =>
        !_resourceDisposed && _state.Status == AsyncResourceStatus.Loading;

    private async Task StartAsync(StartIntent intent, CancellationToken externalToken)
    {
        ResourceOperation operation;
        ResourceOperation? previousOperation;
        bool cleanupDiscarded;
        T? discarded;

        lock (_resourceGate)
        {
            if (_resourceDisposed) return;
            if (externalToken.IsCancellationRequested) return;
            if (intent == StartIntent.Load && _state.Status != AsyncResourceStatus.Idle) return;
            if (intent == StartIntent.Reload && _state.Status == AsyncResourceStatus.Idle) return;

            previousOperation = _operation;
            unchecked { _operationIdentity++; }

            cleanupDiscarded = false;
            discarded = default;
            if (_retention == AsyncResourceRetention.DiscardPrevious &&
                TryGetValue(_stableState, out discarded))
            {
                cleanupDiscarded = true;
                _stableState = AsyncResourceState<T>.Idle();
            }

            var baseline = _stableState;
            var cts = CancellationTokenSource.CreateLinkedTokenSource(externalToken);
            operation = new ResourceOperation(_operationIdentity, baseline, cts);
            _operation = operation;

            _state = _retention == AsyncResourceRetention.RetainPrevious &&
                     TryGetValue(baseline, out var retained)
                ? AsyncResourceState<T>.Loading(retained!)
                : AsyncResourceState<T>.Loading();
        }

        previousOperation?.Cancel();
        if (cleanupDiscarded) Cleanup(discarded!);
        NotifyStateChanged();
        // Register only after Loading has been published. Registration invokes
        // synchronously if cancellation raced admission, so the handler then
        // publishes the restored baseline in the correct visible order.
        operation.Registration = operation.Cancellation.Token.Register(
            () => HandleCancellation(operation));

        if (operation.Cancelled.Task.IsCompleted)
        {
            operation.Dispose();
            return;
        }

        Task<T> loaderTask;
        try
        {
            loaderTask = _loader(operation.Cancellation.Token);
        }
        catch (Exception error)
        {
            loaderTask = Task.FromException<T>(error);
        }
        operation.LoaderTask = loaderTask;

        var completed = await Task.WhenAny(loaderTask, operation.Cancelled.Task)
            .ConfigureAwait(false);
        if (ReferenceEquals(completed, operation.Cancelled.Task))
        {
            ObserveLate(operation, loaderTask);
            return;
        }

        T value;
        try
        {
            value = await loaderTask.ConfigureAwait(false);
        }
        catch (Exception error)
        {
            if (!IsOperationCurrent(operation))
            {
                operation.Dispose();
                return;
            }
            if (error is OperationCanceledException && operation.Cancellation.IsCancellationRequested)
            {
                HandleCancellation(operation);
                operation.Dispose();
                return;
            }
            CompleteFailure(operation, error);
            operation.Dispose();
            return;
        }

        if (!CompleteSuccess(operation, value)) Cleanup(value);
        operation.Dispose();
    }

    private void HandleCancellation(ResourceOperation operation)
    {
        var notify = false;
        lock (_resourceGate)
        {
            operation.Cancelled.TrySetResult(true);
            if (IsCurrentUnsafe(operation))
            {
                unchecked { _operationIdentity++; }
                _operation = null;
                _state = operation.Baseline;
                notify = true;
            }
        }
        if (notify) NotifyStateChanged();
    }

    private bool CompleteSuccess(ResourceOperation operation, T value)
    {
        bool hasPrevious;
        T? previous;
        lock (_resourceGate)
        {
            if (!IsCurrentUnsafe(operation)) return false;
            _operation = null;
            hasPrevious = TryGetValue(_stableState, out previous);
            _stableState = AsyncResourceState<T>.Ready(value);
            _state = _stableState;
        }
        if (hasPrevious) Cleanup(previous!);
        NotifyStateChanged();
        return true;
    }

    private void CompleteFailure(ResourceOperation operation, Exception error)
    {
        lock (_resourceGate)
        {
            if (!IsCurrentUnsafe(operation)) return;
            _operation = null;
            _stableState = _retention == AsyncResourceRetention.RetainPrevious &&
                           TryGetValue(_stableState, out var previous)
                ? AsyncResourceState<T>.Failed(previous!, error)
                : AsyncResourceState<T>.Failed(error);
            _state = _stableState;
        }
        NotifyStateChanged();
    }

    private bool IsOperationCurrent(ResourceOperation operation)
    {
        lock (_resourceGate) return IsCurrentUnsafe(operation);
    }

    private bool IsCurrentUnsafe(ResourceOperation operation) =>
        !_resourceDisposed && _operationIdentity == operation.Identity &&
        ReferenceEquals(_operation, operation);

    private void ObserveLate(ResourceOperation operation, Task<T> loaderTask) =>
        _ = loaderTask.ContinueWith(
            task =>
            {
                try
                {
                    if (task.Status == TaskStatus.RanToCompletion) Cleanup(task.Result);
                    _ = task.Exception;
                }
                finally
                {
                    operation.Dispose();
                }
            },
            CancellationToken.None,
            TaskContinuationOptions.ExecuteSynchronously,
            TaskScheduler.Default);

    private void NotifyStateChanged()
    {
        NotifyPropertyChanged(nameof(State));
        LoadCommand.RaiseCanExecuteChanged();
        ReloadCommand.RaiseCanExecuteChanged();
        CancelCommand.RaiseCanExecuteChanged();
    }

    private void Cleanup(T value)
    {
        try { _cleanupValue?.Invoke(value); }
        catch { /* best-effort ownership cleanup, matching ComponentVMBase.Own */ }
    }

    private static bool TryGetValue(AsyncResourceState<T> state, out T? value)
    {
        value = state.Value;
        return state.HasValue;
    }

    private enum StartIntent
    {
        Load,
        Reload,
    }

    private sealed class ResourceOperation : IDisposable
    {
        public ResourceOperation(
            long identity,
            AsyncResourceState<T> baseline,
            CancellationTokenSource cancellation)
        {
            Identity = identity;
            Baseline = baseline;
            Cancellation = cancellation;
            Cancelled = new TaskCompletionSource<bool>(
                TaskCreationOptions.RunContinuationsAsynchronously);
        }

        public long Identity { get; }
        public AsyncResourceState<T> Baseline { get; }
        public CancellationTokenSource Cancellation { get; }
        public TaskCompletionSource<bool> Cancelled { get; }
        public CancellationTokenRegistration Registration { get; set; }
        public Task<T>? LoaderTask { get; set; }

        public void Cancel()
        {
            try { Cancellation.Cancel(); }
            catch (ObjectDisposedException) { }
            // The token callback restores the stable baseline before the
            // prompt-return sentinel can release StartAsync. Otherwise a fast
            // loader cancellation can dispose this registration first.
            Cancelled.TrySetResult(true);
        }

        public void Dispose()
        {
            Registration.Dispose();
            Cancellation.Dispose();
        }
    }
}

File: langs/python/src/vmx/state/async_resource_vm.py
"""One cancellable asynchronously acquired presentation value.

Spec: spec/23-async-resource-vm.md; ADR-0100.
"""

from __future__ import annotations

import asyncio
import dataclasses
from collections.abc import Awaitable, Callable
from enum import Enum
from typing import Generic, Literal, TypeAlias, TypeVar

from vmx.commands.async_relay_command import AsyncRelayCommand
from vmx.commands.relay_command import RelayCommand
from vmx.components.base import _ComponentVMBase
from vmx.components.protocols import ViewModelType
from vmx.messages.protocols import Message
from vmx.services.dispatcher import Dispatcher
from vmx.services.message_hub import MessageHubProto

T = TypeVar("T")


class AsyncResourceStatus(str, Enum):
    IDLE = "Idle"
    LOADING = "Loading"
    READY = "Ready"
    ERROR = "Error"


class AsyncResourceRetention(str, Enum):
    DISCARD_PREVIOUS = "DiscardPrevious"
    RETAIN_PREVIOUS = "RetainPrevious"


@dataclasses.dataclass(frozen=True)
class AsyncResourceIdle(Generic[T]):
    status: Literal[AsyncResourceStatus.IDLE] = dataclasses.field(
        default=AsyncResourceStatus.IDLE, init=False
    )


@dataclasses.dataclass(frozen=True)
class AsyncResourceLoading(Generic[T]):
    status: Literal[AsyncResourceStatus.LOADING] = dataclasses.field(
        default=AsyncResourceStatus.LOADING, init=False
    )


@dataclasses.dataclass(frozen=True)
class AsyncResourceLoadingWithValue(Generic[T]):
    value: T
    status: Literal[AsyncResourceStatus.LOADING] = dataclasses.field(
        default=AsyncResourceStatus.LOADING, init=False
    )


@dataclasses.dataclass(frozen=True)
class AsyncResourceReady(Generic[T]):
    value: T
    status: Literal[AsyncResourceStatus.READY] = dataclasses.field(
        default=AsyncResourceStatus.READY, init=False
    )


@dataclasses.dataclass(frozen=True)
class AsyncResourceError(Generic[T]):
    error: BaseException
    status: Literal[AsyncResourceStatus.ERROR] = dataclasses.field(
        default=AsyncResourceStatus.ERROR, init=False
    )


@dataclasses.dataclass(frozen=True)
class AsyncResourceErrorWithValue(Generic[T]):
    value: T
    error: BaseException
    status: Literal[AsyncResourceStatus.ERROR] = dataclasses.field(
        default=AsyncResourceStatus.ERROR, init=False
    )


AsyncResourceState: TypeAlias = (
    AsyncResourceIdle[T]
    | AsyncResourceLoading[T]
    | AsyncResourceLoadingWithValue[T]
    | AsyncResourceReady[T]
    | AsyncResourceError[T]
    | AsyncResourceErrorWithValue[T]
)
StableAsyncResourceState: TypeAlias = (
    AsyncResourceIdle[T]
    | AsyncResourceReady[T]
    | AsyncResourceError[T]
    | AsyncResourceErrorWithValue[T]
)


@dataclasses.dataclass(frozen=True)
class _PresentValue(Generic[T]):
    value: T


@dataclasses.dataclass(frozen=True)
class _AbsentValue:
    pass


_ABSENT_VALUE = _AbsentValue()


@dataclasses.dataclass
class _Operation(Generic[T]):
    identity: int
    task: asyncio.Task[T]
    cancelled: asyncio.Future[None]
    baseline: StableAsyncResourceState[T]
    late_cleanup_registered: bool = False


def _value_of(state: StableAsyncResourceState[T]) -> _PresentValue[T] | _AbsentValue:
    if isinstance(state, AsyncResourceReady | AsyncResourceErrorWithValue):
        return _PresentValue(state.value)
    return _ABSENT_VALUE


class AsyncResourceVM(Generic[T], _ComponentVMBase):
    """Component viewmodel for one cancellable asynchronously acquired value."""

    def __init__(
        self,
        *,
        name: str,
        loader: Callable[[], Awaitable[T]],
        hub: MessageHubProto[Message],
        dispatcher: Dispatcher,
        hint: str = "",
        retention: AsyncResourceRetention = AsyncResourceRetention.DISCARD_PREVIOUS,
        cleanup_value: Callable[[T], None] | None = None,
    ) -> None:
        super().__init__(
            name=name,
            hint=hint,
            hub=hub,
            dispatcher=dispatcher,
        )
        self._loader = loader
        self._retention = retention
        self._cleanup_value = cleanup_value
        self._state: AsyncResourceState[T] = AsyncResourceIdle()
        self._stable_state: StableAsyncResourceState[T] = AsyncResourceIdle()
        self._operation_identity = 0
        self._operation: _Operation[T] | None = None
        self._resource_disposed = False

        self._load_command = (
            AsyncRelayCommand.builder().task(self.load).predicate(self._can_load).build()
        )
        self._reload_command = (
            AsyncRelayCommand.builder().task(self.reload).predicate(self._can_reload).build()
        )
        self._cancel_command = (
            RelayCommand.builder().task(self.cancel).predicate(self._can_cancel).build()
        )

    @property
    def type(self) -> ViewModelType:
        return ViewModelType.COMPONENT

    @property
    def state(self) -> AsyncResourceState[T]:
        return self._state

    @property
    def load_command(self) -> AsyncRelayCommand:
        return self._load_command

    @property
    def reload_command(self) -> AsyncRelayCommand:
        return self._reload_command

    @property
    def cancel_command(self) -> RelayCommand:
        return self._cancel_command

    async def load(self) -> None:
        if not self._can_load():
            return
        await self._start()

    async def reload(self) -> None:
        if not self._can_reload():
            return
        await self._start()

    def cancel(self) -> None:
        operation = self._operation
        if not self._can_cancel() or operation is None:
            return
        self._operation_identity += 1
        self._operation = None
        self._cancel_operation(operation)
        self._set_state(operation.baseline)
        self._load_command.cancel()
        self._reload_command.cancel()

    def _can_load(self) -> bool:
        return not self._resource_disposed and self._state.status is AsyncResourceStatus.IDLE

    def _can_reload(self) -> bool:
        return not self._resource_disposed and self._state.status is not AsyncResourceStatus.IDLE

    def _can_cancel(self) -> bool:
        return not self._resource_disposed and self._state.status is AsyncResourceStatus.LOADING

    async def _start(self) -> None:
        previous_operation = self._operation
        self._operation_identity += 1
        identity = self._operation_identity

        if self._retention is AsyncResourceRetention.DISCARD_PREVIOUS:
            previous = _value_of(self._stable_state)
            if isinstance(previous, _PresentValue):
                self._stable_state = AsyncResourceIdle()
                self._cleanup(previous.value)

        # Cleanup is user code and may dispose the VM or start a newer intent.
        # Do not create (and therefore invoke) a loader after either terminal
        # transition has superseded this start.
        if self._resource_disposed or self._operation_identity != identity:
            return

        baseline = self._stable_state
        retained = (
            _value_of(baseline)
            if self._retention is AsyncResourceRetention.RETAIN_PREVIOUS
            else _ABSENT_VALUE
        )
        loading: AsyncResourceState[T]
        if isinstance(retained, _PresentValue):
            loading = AsyncResourceLoadingWithValue(retained.value)
        else:
            loading = AsyncResourceLoading()

        async def invoke_loader() -> T:
            return await self._loader()

        task = asyncio.create_task(invoke_loader())
        cancelled: asyncio.Future[None] = asyncio.get_running_loop().create_future()
        operation = _Operation(identity, task, cancelled, baseline)
        self._operation = operation
        if previous_operation is not None:
            self._cancel_operation(previous_operation)
        self._set_state(loading)

        try:
            done, _pending = await asyncio.wait(
                (task, cancelled), return_when=asyncio.FIRST_COMPLETED
            )
        except asyncio.CancelledError:
            if self._is_operation_current(operation):
                self._operation_identity += 1
                self._operation = None
                self._cancel_operation(operation)
                self._set_state(baseline)
            self._register_late_cleanup(operation)
            raise

        if cancelled in done:
            self._register_late_cleanup(operation)
            return

        try:
            value = task.result()
        except asyncio.CancelledError:
            return
        except BaseException as error:
            if not self._is_operation_current(operation):
                return
            self._operation = None
            if self._retention is AsyncResourceRetention.RETAIN_PREVIOUS:
                previous = _value_of(self._stable_state)
            else:
                previous = _ABSENT_VALUE
            failed: StableAsyncResourceState[T]
            if isinstance(previous, _PresentValue):
                failed = AsyncResourceErrorWithValue(previous.value, error)
            else:
                failed = AsyncResourceError(error)
            self._stable_state = failed
            self._set_state(failed)
            return

        if not self._is_operation_current(operation):
            self._cleanup(value)
            return

        self._operation = None
        previous = _value_of(self._stable_state)
        ready: StableAsyncResourceState[T] = AsyncResourceReady(value)
        self._stable_state = ready
        if isinstance(previous, _PresentValue):
            self._cleanup(previous.value)
        self._set_state(ready)

    def _is_operation_current(self, operation: _Operation[T]) -> bool:
        return (
            not self._resource_disposed
            and self._operation_identity == operation.identity
            and self._operation is operation
        )

    @staticmethod
    def _cancel_operation(operation: _Operation[T]) -> None:
        if not operation.cancelled.done():
            operation.cancelled.set_result(None)
        if not operation.task.done():
            operation.task.cancel()

    def _cleanup_late_task(self, task: asyncio.Task[T]) -> None:
        try:
            value = task.result()
        except BaseException:
            return
        self._cleanup(value)

    def _register_late_cleanup(self, operation: _Operation[T]) -> None:
        if operation.late_cleanup_registered:
            return
        operation.late_cleanup_registered = True
        operation.task.add_done_callback(self._cleanup_late_task)

    def _set_state(self, state: AsyncResourceState[T]) -> None:
        if self._resource_disposed or self._state is state:
            return
        self._state = state
        self._notify_property_changed("state")
        self._load_command.raise_can_execute_changed()
        self._reload_command.raise_can_execute_changed()
        self._cancel_command.raise_can_execute_changed()

    def _cleanup(self, value: T) -> None:
        if self._cleanup_value is None:
            return
        try:
            self._cleanup_value(value)
        except BaseException:
            pass

    def _on_dispose(self) -> None:
        if self._resource_disposed:
            return
        self._resource_disposed = True
        self._operation_identity += 1
        operation = self._operation
        self._operation = None
        if operation is not None:
            self._cancel_operation(operation)
            self._register_late_cleanup(operation)
        self._load_command.cancel()
        self._reload_command.cancel()
        self._load_command.dispose()
        self._reload_command.dispose()
        self._cancel_command.dispose()
        accepted = _value_of(self._stable_state)
        self._stable_state = AsyncResourceIdle()
        if isinstance(accepted, _PresentValue):
            self._cleanup(accepted.value)
